File: 8. String to Integer (atoi)/StringtoInteger.java
class Solution {
    public int myAtoi(String str) {
        int sign = 0;
        int num = 0;
        for (int i = 0; i < str.length(); ++i) {
            if (sign == 0 && str.charAt(i) == ' ') {
                continue;
            } else {
                if (sign == 0 && str.charAt(i) == '+') {
                    sign = 1;
                } else if (sign == 0 && str.charAt(i) == '-') {
                    sign = -1;
                } else if (str.charAt(i) >= '0' && str.charAt(i) <= '9') {
                    int n = Integer.parseInt(Character.toString(str.charAt(i)));
                    if (sign == 0) sign = 1;
                    if (num > Integer.MAX_VALUE / 10 
                        || (num == Integer.MAX_VALUE / 10 && n > 7)
                        || num < Integer.MIN_VALUE / 10
                        || (num == Integer.MIN_VALUE / 10 && n < -8)) {
                        if (sign == 1) {
                            return Integer.MAX_VALUE;
                        } else {
                            return Integer.MIN_VALUE;
                        }
                    }
                    num = num * 10 + n;
                } else {
                    return sign * num;
                }
            }
        }
        return sign * num;
    }
}

File: 8. String to Integer (atoi)/StringtoInteger.js
/**
 * @param {string} str
 * @return {number}
 */
var myAtoi = function(str) {
    var sign = 0;
    var num = 0;
    for (var i = 0; i < str.length; ++i) {
        if (sign == 0 && str[i] == ' ') {
            continue;
        } else {
            if (sign == 0 && str[i] == '+') {
                sign = 1;
            } else if (sign == 0 && str[i] == '-') {
                sign = -1;
            } else if (str[i] >= '0' && str[i] <= '9') {
                var n = parseInt(str[i]);
                if (sign == 0) sign = 1;
                num = num * 10 + n;
                if (num > 2147483647 && sign == 1) return 2147483647;
                if (num > 2147483647 && sign == -1) return -2147483648;
            } else {
                return sign * num;
            }
        }
    }
    return sign * num;
};
